### PR TITLE
[aptos-rosetta] Ensure construction APIs all work

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Some inconsistent inputs, and additional missing debug data was making it hard to test.  Now, there's proper debug data, the DSL is complete, with a possible need  of a few tweaks, and there may be some work around some edge cases left.

I will be coming back around to fix documentation, and some of the other edge cases after this PR.

### Test Plan
```
[STATS] Blocks: 1225 (Orphaned: 0) Transactions: 2483 Operations: 119 Accounts: 17 Reconciliations: 117 (Inactive: 46, Exempt: 0, Skipped: 0, Cov)

Success: Reconciliation Coverage End Condition [Coverage: 100.000000%]

+--------------------+--------------------------------+--------+
|  CHECK:DATA TESTS  |          DESCRIPTION           | STATUS |
+--------------------+--------------------------------+--------+
| Request/Response   | Rosetta implementation         | PASSED |
|                    | serviced all requests          |        |
+--------------------+--------------------------------+--------+
| Response Assertion | All responses are correctly    | PASSED |
|                    | formatted                      |        |
+--------------------+--------------------------------+--------+
| Block Syncing      | Blocks are connected into a    | PASSED |
|                    | single canonical chain         |        |
+--------------------+--------------------------------+--------+
| Balance Tracking   | Account balances did not go    | PASSED |
|                    | negative                       |        |
+--------------------+--------------------------------+--------+
| Reconciliation     | No balance discrepancies were  | PASSED |
|                    | found between computed and     |        |
|                    | live balances                  |        |
+--------------------+--------------------------------+--------+

+--------------------------+--------------------------------+-------------+
|     CHECK:DATA STATS     |          DESCRIPTION           |    VALUE    |
+--------------------------+--------------------------------+-------------+
| Blocks                   | # of blocks synced             |        1227 |
+--------------------------+--------------------------------+-------------+
| Orphans                  | # of blocks orphaned           |           0 |
+--------------------------+--------------------------------+-------------+
| Transactions             | # of transaction processed     |        2487 |
+--------------------------+--------------------------------+-------------+
| Operations               | # of operations processed      |         119 |
+--------------------------+--------------------------------+-------------+
| Accounts                 | # of accounts seen             |          17 |
+--------------------------+--------------------------------+-------------+
| Active Reconciliations   | # of reconciliations performed |          71 |
|                          | after seeing an account in a   |             |
|                          | block                          |             |
+--------------------------+--------------------------------+-------------+
| Inactive Reconciliations | # of reconciliations performed |          46 |
|                          | on randomly selected accounts  |             |
+--------------------------+--------------------------------+-------------+
| Exempt Reconciliations   | # of reconciliation failures   |           0 |
|                          | considered exempt              |             |
+--------------------------+--------------------------------+-------------+
| Failed Reconciliations   | # of reconciliation failures   |           0 |
+--------------------------+--------------------------------+-------------+
| Skipped Reconciliations  | # of reconciliations skipped   |           0 |
+--------------------------+--------------------------------+-------------+
| Reconciliation Coverage  | % of accounts that have been   | 100.000000% |
|                          | reconciled                     |             |
+--------------------------+--------------------------------+-------------+

badger 2022/07/13 12:37:15 INFO: Storing value log head: {Fid:0 Len:32 Offset:1856309}



[STATS] Transactions Confirmed: 39 (Created: 45, In Progress: 6, Stale: 0, Failed: 0) Addresses Created: 19

Success: {"create_account":10,"transfer":20}

+--------------------------+--------------------------------+-------+
| CHECK:CONSTRUCTION STATS |          DESCRIPTION           | VALUE |
+--------------------------+--------------------------------+-------+
| Addresses Created        | # of addresses created         |    19 |
+--------------------------+--------------------------------+-------+
| Transactions Created     | # of transactions created      |    45 |
+--------------------------+--------------------------------+-------+
| Stale Broadcasts         | # of broadcasts missing after  |     0 |
|                          | stale depth                    |       |
+--------------------------+--------------------------------+-------+
| Transactions Confirmed   | # of transactions seen         |    39 |
|                          | on-chain                       |       |
+--------------------------+--------------------------------+-------+
| Failed Broadcasts        | # of transactions that         |     0 |
|                          | exceeded broadcast limit       |       |
+--------------------------+--------------------------------+-------+
+------------------------------+-------+
| CHECK:CONSTRUCTION WORKFLOWS | COUNT |
+------------------------------+-------+
| create_account               |    19 |
+------------------------------+-------+
| request_funds                |     0 |
+------------------------------+-------+
| transfer                     |    20 |
+------------------------------+-------+


```

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1931)

### CODE REVIEW

File: crates/aptos-rosetta/src/block.rs [hunk 1]
```diff
@@ -166,30 +172,73 @@ impl BlockCache {
...
+                }
+            });
+
+            let search_right = map.iter().find_map(|(i, info)| {
+                if *i > block_index {
+                    Some((i, info, block_index - i))
+                } else {
+                    None
+                }
+            });
```
Comment (davidiw) [inline crates/aptos-rosetta/src/block.rs:198]: feels like you could probably do this in a single statement
```
let (search_left, search_right) = ... {
  if ... {
    (None, Some)
  } else if {
    (Some, None)
  } else {
    None
  }
});
```

File: crates/aptos-rosetta/src/block.rs [hunk 2]
```diff
@@ -166,30 +172,73 @@ impl BlockCache {
...
+                }
+                (None, Some((i, info, _))) => {
+                    let start = info.start_version - 1;
+                    let range = (i - 1)..=block_index;
+                    (start, range, false)
+                }
+                (
+                    Some((left_index, left_info, left_distance)),
+                    Some((right_index, right_info, right_distance)),
+                ) => {
```
Comment (davidiw) [inline crates/aptos-rosetta/src/block.rs:214]: how is this case even possible?

Comment (gregnazario) [reply crates/aptos-rosetta/src/block.rs:214]: You're searching the cache, and if there's nothing in the cache.  But I seed the cache when I first create it.

File: crates/aptos-rosetta/src/block.rs (hunk 1 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/block.rs:198]: Sure, I can do that

Comment (gregnazario) [reply crates/aptos-rosetta/src/block.rs:198]: I just did it this way for readability

**Review by davidiw — APPROVED** [review]

**Pull request merged**